File: channel-manager/frontend/src/main/java/org/onehippo/cms7/channelmanager/RootPanel.java
/*
 *  Copyright 2011-2017 Hippo B.V. (http://www.onehippo.com)
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.onehippo.cms7.channelmanager;

import java.util.Map;

import org.apache.wicket.markup.head.IHeaderResponse;
import org.apache.wicket.model.Model;
import org.apache.wicket.request.cycle.RequestCycle;
import org.apache.wicket.request.handler.resource.ResourceReferenceRequestHandler;
import org.apache.wicket.request.resource.PackageResourceReference;
import org.hippoecm.frontend.PluginRequestTarget;
import org.hippoecm.frontend.plugin.IPluginContext;
import org.hippoecm.frontend.plugin.config.IPluginConfig;
import org.hippoecm.frontend.plugins.yui.layout.IWireframe;
import org.hippoecm.frontend.plugins.yui.layout.WireframeUtils;
import org.hippoecm.frontend.service.IRestProxyService;
import org.json.JSONException;
import org.json.JSONObject;
import org.onehippo.cms7.channelmanager.channeleditor.ChannelEditor;
import org.onehippo.cms7.channelmanager.channels.BlueprintStore;
import org.onehippo.cms7.channelmanager.channels.ChannelOverview;
import org.onehippo.cms7.channelmanager.channels.ChannelStore;
import org.onehippo.cms7.channelmanager.channels.ChannelStoreFactory;
import org.onehippo.cms7.channelmanager.restproxy.RestProxyServicesManager;
import org.onehippo.cms7.channelmanager.widgets.ExtLinkPicker;
import org.wicketstuff.js.ext.ExtPanel;
import org.wicketstuff.js.ext.layout.BorderLayout;
import org.wicketstuff.js.ext.util.ExtClass;
import org.wicketstuff.js.ext.util.ExtProperty;
import org.wicketstuff.js.ext.util.JSONIdentifier;

@ExtClass("Hippo.ChannelManager.RootPanel")
public class RootPanel extends ExtPanel {

    private static final long serialVersionUID = 1L;

    private static final PackageResourceReference BREADCRUMB_ARROW = new PackageResourceReference(RootPanel.class, "breadcrumb-arrow.png");

    public enum CardId {
        CHANNEL_MANAGER(0),
        CHANNEL_EDITOR(1);

        private final int tabIndex;

        CardId(int tabIndex) {
            this.tabIndex = tabIndex;
        }

        int getTabIndex() {
            return tabIndex;
        }

    }

    public static final String CONFIG_CHANNEL_LIST = "channel-list";
    public static final String CONFIG_TEMPLATE_COMPOSER = "templatecomposer";
    public static final String COMPOSER_REST_MOUNT_PATH_PROPERTY = "composerRestMountPath";
    public static final String DEFAULT_COMPOSER_REST_MOUNT_PATH = "/_rp";

    private BlueprintStore blueprintStore;
    private ChannelStore channelStore;
    private ChannelEditor channelEditor;
    private ExtStoreFuture<Object> channelStoreFuture;

    private boolean redraw = false;

    @ExtProperty
    private final String perspectiveId;

    @ExtProperty
    private int activeItem = 0;

    @ExtProperty
    @SuppressWarnings("unused")
    private String composerRestMountPath;

    @ExtProperty
    private final String[] contextPaths;

    @ExtProperty
    private boolean showBreadcrumbInitially = false;

    @Override
    public void buildInstantiationJs(final StringBuilder js, final String extClass, final JSONObject properties) {
        js.append("try { ");
        super.buildInstantiationJs(js, extClass, properties);
        js.append("} catch(exception) { console.log('Error initializing channel manager. '+exception); } ");
    }

    public RootPanel(final IPluginContext context, final IPluginConfig config, final String id, final String perspectiveId) {
        super(id);

        this.perspectiveId = perspectiveId;

        final IPluginConfig channelListConfig = config.getPluginConfig(CONFIG_CHANNEL_LIST);

        // card 1: template composer
        final IPluginConfig editorConfig = config.getPluginConfig(CONFIG_TEMPLATE_COMPOSER);
        if (editorConfig == null) {
            composerRestMountPath = DEFAULT_COMPOSER_REST_MOUNT_PATH;
        } else {
            composerRestMountPath = editorConfig.getString(COMPOSER_REST_MOUNT_PATH_PROPERTY, DEFAULT_COMPOSER_REST_MOUNT_PATH);
        }

        final Map<String, IRestProxyService> liveRestProxyServices = RestProxyServicesManager.getLiveRestProxyServices(context, config);
        contextPaths = liveRestProxyServices.keySet().toArray(new String[liveRestProxyServices.size()]);

        this.blueprintStore = new BlueprintStore(liveRestProxyServices);
        this.channelStore = ChannelStoreFactory.createStore(context, channelListConfig, liveRestProxyServices, blueprintStore);
        this.channelStoreFuture = new ExtStoreFuture<>(channelStore);
        add(this.channelStore);
        add(this.channelStoreFuture);


        // channel manager
        final ExtPanel channelManagerCard = new ExtPanel();
        channelManagerCard.setBorder(false);
        channelManagerCard.setTitle(Model.of(getString("channel-manager")));
        channelManagerCard.setHeader(false);
        channelManagerCard.setLayout(new BorderLayout());

        final ChannelOverview channelOverview = new ChannelOverview(channelListConfig, composerRestMountPath,
                this.channelStoreFuture, !blueprintStore.isEmpty());
        channelOverview.setRegion(BorderLayout.Region.CENTER);
        channelManagerCard.add(channelOverview);

        channelManagerCard.add(this.blueprintStore);
        add(channelManagerCard);

        // channel editor
        channelEditor = new ChannelEditor(context, editorConfig, composerRestMountPath, channelStoreFuture, contextPaths);
        add(channelEditor);

        // folder picker
        add(new ExtLinkPicker(context));
    }

    public void redraw() {
        redraw = true;
    }

    public void render(final PluginRequestTarget target) {
        channelStore.update();
        if (target != null) {
            if (redraw) {
                selectActiveItem(target);
                redraw = false;
            }
        } else {
            this.showBreadcrumbInitially = true;
        }
    }

    private void selectActiveItem(final PluginRequestTarget target) {
        final String script = String.format("Ext.getCmp('rootPanel').selectCard(%s);", activeItem);
        target.appendJavaScript(script);
    }

    @Override
    public void renderHead(final IHeaderResponse response) {
        final IWireframe parentWireframe = WireframeUtils.getParentWireframe(this);
        if (parentWireframe != null) {
            response.render(parentWireframe.getHeaderItem());
        }

        super.renderHead(response);

        response.render(ChannelManagerHeaderItem.get());
    }

    @Override
    protected void preRenderExtHead(StringBuilder js) {
        blueprintStore.onRenderExtHead(js);
        channelStore.onRenderExtHead(js);
        channelStoreFuture.onRenderExtHead(js);
        super.preRenderExtHead(js);
    }

    @Override
    protected void onRenderProperties(JSONObject properties) throws JSONException {
        super.onRenderProperties(properties);
        properties.put("blueprintStore", new JSONIdentifier(this.blueprintStore.getJsObjectId()));
        properties.put("channelStore", new JSONIdentifier(this.channelStore.getJsObjectId()));
        properties.put("channelStoreFuture", new JSONIdentifier(this.channelStoreFuture.getJsObjectId()));

        RequestCycle rc = RequestCycle.get();
        properties.put("breadcrumbIconUrl", rc.urlFor(new ResourceReferenceRequestHandler(
                BREADCRUMB_ARROW)));
    }

    public ChannelEditor getChannelEditor() {
        return this.channelEditor;
    }

    public void setActiveCard(CardId rootPanelCard) {
        this.activeItem = rootPanelCard.getTabIndex();
        redraw();
    }

}

File: channel-manager/frontend/src/main/java/org/onehippo/cms7/channelmanager/RootPanel.js
/**
 * Copyright 2011-2013 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the  "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
(function() {
    "use strict";

    Ext.namespace('Hippo.ChannelManager');

    /**
     * @class Hippo.ChannelManager.RootPanel
     * @extends Ext.Panel
     */
    Hippo.ChannelManager.RootPanel = Ext.extend(Ext.Panel, {

        constructor: function(config) {
            this.channelStore = config.channelStore;
            this.blueprintStore = config.blueprintStore;
            this.resources = config.resources;
            this.selectedChannelId = null;

            this.toolbar = new Hippo.ChannelManager.BreadcrumbToolbar({
                autoWidth: true,
                id: 'hippo-channelmanager-breadcrumb',
                breadcrumbIconUrl: config.breadcrumbIconUrl,
                layoutConfig: {
                    pack: 'left'
                }
            });

            Ext.apply(config, {
                id: 'rootPanel',
                layout: 'card',
                layoutOnCardChange: true,
                deferredRender: true,
                viewConfig: {
                    forceFit: true
                },
                border: false,
                tbar: this.toolbar
            });

            Hippo.ChannelManager.RootPanel.superclass.constructor.call(this, config);

            this.selectCard(config.activeItem);
        },

        initComponent: function() {
            var self, channelSelectedHandler;

            self = this;

            // recalculate the ExtJs layout when the YUI layout manager fires a resize event
            this.on('afterlayout', function() {
                var yuiLayout = this.getEl().findParent("div.yui-layout-unit");
                YAHOO.hippo.LayoutManager.registerResizeListener(yuiLayout, this, function(sizes) {
                    self.setSize(sizes.body.w, sizes.body.h);
                    self.doLayout();
                }, true);
            }, this, {single: true});

            // get all child components
            this.win = new Hippo.ChannelManager.NewChannelWindow({
                blueprintStore: self.blueprintStore,
                channelStore: self.channelStore,
                resources: self.resources
            });
            this.formPanel = Ext.getCmp('channel-form-panel');
            this.gridPanel = Ext.getCmp('channel-grid-panel');
            this.channelIconPanel = Ext.getCmp('channelIconPanel');
            this.channelOverviewPanel = Ext.getCmp('channelOverview');

            // register channel creation events
            this.channelOverviewPanel.on('add-channel', function() {
                this.win.show();
            }, this);
            this.formPanel.on('channel-created', function() {
                this.win.hide();
                this.channelStore.reload();
            }, this);

            channelSelectedHandler = function(channelId, record) {
                this.selectedChannelId = channelId;
                // don't activate template composer when it is already active
                if (this.layout.activeItem === Hippo.ChannelManager.TemplateComposer.Instance) {
                    return;
                }
                Hippo.ChannelManager.TemplateComposer.Instance.browseTo({ channelId: channelId });
                Ext.getCmp('rootPanel').showTemplateComposer();
                // TODO fix, I have no clue why the template composer card is not getting activated properly
                document.getElementById('Hippo.ChannelManager.TemplateComposer.Instance').className = 'x-panel';
            };

            this.gridPanel.on('channel-selected', channelSelectedHandler, this);
            this.channelIconPanel.on('channel-selected', channelSelectedHandler, this);

            Hippo.ChannelManager.TemplateComposer.Instance.on('mountChanged', function(data) {
                var channelRecord = this.gridPanel.getChannelByMountId(data.mountId),
                        firstChange = data.oldMountId === null;
                if (!firstChange && this.selectedChannelId !== channelRecord.get('id')) {
                    this.gridPanel.selectChannel(channelRecord.get('id'));
                }
            }, this);

            Hippo.ChannelManager.RootPanel.superclass.initComponent.apply(this, arguments);
        },

        selectCard: function(itemId) {
            while (this.toolbar.getBreadcrumbSize() > 0) {
                this.toolbar.popItem();
            }

            this.toolbar.pushItem({
                card: this.items.get(0),
                click: function() {
                    this.layout.setActiveItem(0);
                },
                scope: this
            });
            if ((typeof itemId !== 'undefined') && itemId !== 0) {
                this.toolbar.pushItem({
                    card: this.items.get(itemId),
                    click: function() {
                        this.layout.setActiveItem(itemId);
                    },
                    scope: this
                });

                this.layout.setActiveItem(itemId);
                if (itemId === 1) {
                    // TODO fix, I have no clue why the template composer card is not getting activated properly
                    document.getElementById('Hippo.ChannelManager.TemplateComposer.Instance').className = 'x-panel';
                }
            } else {
                this.layout.setActiveItem(0);
            }
        },

        update: function(config) {
            this.selectCard(config.activeItem);
        },

        showChannelManager: function() {
            this.layout.setActiveItem(0);
        },

        showTemplateComposer: function() {
            this.toolbar.pushItem({
                card: this.items.get(1),
                click: function() {
                    this.layout.setActiveItem(1);
                },
                scope: this
            });
            this.layout.setActiveItem(1);
        },

        showConfigEditor: function() {
            this.toolbar.pushItem({
                card: this.items.get(2),
                click: function() {
                    this.layout.setActiveItem(2);
                },
                scope: this
            });
            this.layout.setActiveItem(2);
        }
    });

    Ext.reg('Hippo.ChannelManager.RootPanel', Hippo.ChannelManager.RootPanel);


    Hippo.ChannelManager.NewChannelWindow = Ext.extend(Ext.Window, {

        constructor: function(config) {
            this.blueprintStore = config.blueprintStore;
            this.channelStore = config.channelStore;
            this.resources = config.resources;

            Hippo.ChannelManager.NewChannelWindow.superclass.constructor.call(this, config);
        },

        initComponent: function() {
            var config = {
                title: this.resources['new-channel-blueprint'],
                width: 720,
                height: 450,
                modal: true,
                resizable: false,
                closeAction: 'hide',
                layout: 'fit',
                items: [
                    {
                        id: 'card-container',
                        layout: 'card',
                        activeItem: 0,
                        layoutConfig: {
                            hideMode: 'offsets',
                            deferredRender: true,
                            layoutOnCardChange: true
                        }
                    }
                ],
                buttonAlign: 'left',
                buttons: [
                    {
                        id: 'previousButton',
                        text: this.resources['new-channel-previous'],
                        handler: this.processPreviousStep,
                        scope: this,
                        hidden: true
                    },
                    '->',
                    {
                        id: 'nextButton',
                        text: this.resources['new-channel-next'],
                        handler: this.processNextStep,
                        scope: this
                    },
                    {
                        id: 'cancelButton',
                        text: this.resources['new-channel-cancel'],
                        scope: this,
                        handler: function() {
                            this.hide();
                        }

                    }
                ]

            };

            Ext.apply(this, Ext.apply(this.initialConfig, config));

            Hippo.ChannelManager.NewChannelWindow.superclass.initComponent.apply(this, arguments);

            this.on('beforeshow', this.resetWizard, this);

            Ext.getCmp('card-container').add(new Hippo.ChannelManager.BlueprintListPanel({
                id: 'blueprints-panel',
                store: this.blueprintStore,
                resources: this.resources
            }));

            Ext.getCmp('card-container').add(new Hippo.ChannelManager.ChannelFormPanel({
                id: 'channel-form-panel',
                store: this.channelStore,
                resources: this.resources
            }));

        },

        showBlueprintChoice: function() {
            this.setTitle(this.resources['new-channel-blueprint']);
            Ext.getCmp('card-container').layout.setActiveItem('blueprints-panel');
            Ext.getCmp('nextButton').setText(this.resources['new-channel-next']);
            Ext.getCmp('previousButton').hide();
        },

        showChannelForm: function() {
            this.setTitle(this.resources['new-channel-properties']);
            Ext.getCmp('card-container').layout.setActiveItem('channel-form-panel');
            Ext.getCmp('nextButton').setText(this.resources['new-channel-create']);
            Ext.getCmp('previousButton').show();
        },

        resetWizard: function() {
            this.showBlueprintChoice();
        },

        processPreviousStep: function() {
            if (Ext.getCmp('card-container').layout.activeItem.id === 'channel-form-panel') {
                this.showBlueprintChoice();
            }
        },

        processNextStep: function() {
            if (Ext.getCmp('card-container').layout.activeItem.id === 'blueprints-panel') {
                this.showChannelForm();
            } else {
                this.submitChannelForm();
            }
        },

        submitChannelForm: function() {
            var nextButton = Ext.getCmp('nextButton');
            nextButton.disable();
            Ext.getCmp('channel-form-panel').submitForm().when(function() {
                nextButton.enable();
            }).otherwise(function() {
                nextButton.enable();
            });
        }

    });

}());
